### DISCUSSION TITLE
ChexkBoxFormatter

### DISCUSSION THREAD

**Discussion by havvanur160401046** [serenity-is/Serenity#5768] [Help]
![image](https://user-images.githubusercontent.com/24252615/120471456-6703e880-c3ad-11eb-8630-ce8cb118538d.png)
![image](https://user-images.githubusercontent.com/24252615/120471968-07f2a380-c3ae-11eb-8991-4932d835b3d6.png)
I also want to change the checkbox over the list, I need to write a checkboxformatter function, but I have no idea how to do it, there is an example for the texts above, I think it should be like this as far as I understand
![image](https://user-images.githubusercontent.com/24252615/120472274-5b64f180-c3ae-11eb-86c1-a73f57a0c3b5.png)
I did it like this, but what are my mistakes or deficiencies?

**Comment by edwardch** [comment]
I'm not sure what you're trying to achieve here?

Try search for `GridRowSelectionMixin` or inspect the checkbox element in your browser to find out more about it.

**Reply by havvanur160401046** [reply]
I want to change the discontunued value. There is an example for string or number on the list, but how can I do it for boolean values?

**Reply by havvanur160401046** [reply]
so I want to make it changeable via the list when it lists as above without opening the form

**Reply by edwardch** [reply]
I haven't tried this in a grid, but if you open a dialog and inspect the checkbox element you will get something like this
```html
<input type="checkbox" class="editor s-BooleanEditor" id="StartSharp_Northwind_ProductDialog39_Discontinued" name="Discontinued">
```

Perhaps try using those classes instead and see how it goes?

**Comment by reach2rv** [comment]
@havvanur160401046 
In your Grid.ts

 ```TypeScript
       private booleantFormatter(ctx) {
            var klass = 'edit check-box';
            var item = ctx.item as ConfigurationRow;
            var pending = this.pendingChanges[item.Id];

            if (pending && pending[ctx.column.field] !== undefined) {
                klass += ' dirty';
            }

            var value = this.getEffectiveValue(item, ctx.column.field) as string;
            var check: string;
            var isChecked = q.ToBool(value);
            if (isChecked)
                check = "<input type='checkbox' checked='" + isChecked + "' class='" + klass +
                    "' data-field='" + ctx.column.field +
                    "' value='" + value + "'/>";
            else
                check = "<input type='checkbox' class='" + klass +
                    "' data-field='" + ctx.column.field +
                    "' value='" + value + "'/>";


            return check
        }
```

``` TypeScript
protected getColumns() {
            var columns = super.getColumns();

            let xyzColumn = Q.first(columns, x => x.field == fld.xyz);

            xyzColumn .format = ctx =>             
                    return this.booleantFormatter(ctx);

            return columns;
        }
```
``` TypeScript
   private inputsChange(e: JQueryEventObject) {
            var cell = this.slickGrid.getCellFromEvent(e);
            var item = this.itemAt(cell.row);
            var input = $(e.target);
            var field = input.data('field');
            var text = Q.coalesce(Q.trimToNull(input.val()), '0');           
            var pending = this.pendingChanges[item.Id];

            var effective = this.getEffectiveValue(item, field);
            var oldText: string; 
            var value;           

            if (!pending) {
                this.pendingChanges[item.Id] = pending = {};
            }
            if (input.hasClass('check-box')) {
                    if (item.Value == "False") {
                        value = "True";
                        input.prop("checked");
                    }
                    else {
                        value = "False";
                        input.prop("");
                    }
             }
            
            pending[field] = value;
            item[field] = value;
            this.view.updateItem(item.Id, item);
            this.view.refresh();

            input.val(value).addClass('dirty');

        }
```


in Constructor 

```TypeScript
this.slickContainer.on('change', '.edit:input', (e) => this.inputsChange(e));
```

I have trimmed down the code by removing other validation for my specific code.. Please try and let me know.. you may need to alter some functions as per your needs

**Reply by reach2rv** [reply]
for q.ToBool I am using serenity extra by @dfaruque if you don't want to add additional package use following function 

``` TypeScript
export function ToBool(value) {
        let val: boolean = false;
        if (value == 'true' || value == 'True')
            val = true;

        return value == 'true' || value == 'True' ? true : false;
    } ```

**Reply by havvanur160401046** [reply]
![image](https://user-images.githubusercontent.com/24252615/120852248-a172ae80-c582-11eb-9404-2448f6f945dd.png)
I got what I wanted but I'm having trouble saving it
the save-changes button doesn't work when I tick it

**Reply by havvanur160401046** [reply]
``    private setSaveButtonState() {
            this.toolbar.findButton('apply-changes-button').toggleClass('disabled',
                Object.keys(this.pendingChanges).length === 0);
        }

        private saveClick() {
            if (Object.keys(this.pendingChanges).length === 0) {
                return;
            }

            // this calls save service for all modified rows, one by one
            // you could write a batch update service
            var keys = Object.keys(this.pendingChanges);
            var current = -1;
            var self = this;

            (function saveNext() {
                if (++current >= keys.length) {
                    self.refresh();
                    return;
                }

                var key = keys[current];
                var entity = Q.deepClone(self.pendingChanges[key]);
                entity.SiparisId = key;
                Q.serviceRequest('Islemler/SiparisYonet/Update', {
                    EntityId: key,
                    Entity: entity
                }, (response) => {
                    delete self.pendingChanges[key];
                    saveNext();
                });
            })();
        }
        private inputsChange(e: JQueryEventObject) {
            var cell = this.slickGrid.getCellFromEvent(e);
            var item = this.itemAt(cell.row);
            var input = $(e.target);
            var field = input.data('field');
            var text = Q.coalesce(Q.trimToNull(input.val()), '0');
            var pending = this.pendingChanges[item.SiparisId];

            var effective = this.getEffectiveValue(item, field);
            var oldText: string;
            var value;

            if (!pending) {
                this.pendingChanges[item.SiparisId] = pending = {};
            }
            if (input.hasClass('check-box')) {
                if (item.Onay.valueOf.toString() == "False") {
                    value = "True";
                    input.prop("checked");
                    return;
                }
                else {
                    value = "False";
                    input.prop("");
                    return;
                }
            }

            pending[field] = value;
            item[field] = value;
            this.view.updateItem(item.Onay, item);
            this.view.refresh();

            input.val(value).addClass('dirty');
            this.setSaveButtonState();

        }

**Reply by havvanur160401046** [reply]
`   protected getButtons() {
            var buttons = super.getButtons();

            buttons.push({
                title: 'Save Changes',
                cssClass: 'apply-changes-button disabled',
                onClick: e => this.saveClick(),
                separator: true
            });

            return buttons;
        }
        protected onViewProcessData(response) {
            this.pendingChanges = {};
            this.setSaveButtonState();
            return super.onViewProcessData(response);
        }
`

**Reply by eissa1005** [reply]
call this on Constructor
this.slickContainer.on('change', '.edit:input', (e) => this.inputsChange(e));

**Reply by reach2rv** [reply]
@havvanur160401046  you are missing this

```this.slickContainer.on('change', '.edit:input', (e) => this.inputsChange(e));```

add it in grid constructor

**Reply by reach2rv** [reply]
@havvanur160401046  Please mark as answered if your issue is resolved. Thanks

**Reply by havvanur160401046** [reply]
![image](https://user-images.githubusercontent.com/24252615/120920352-020b0400-c6c7-11eb-941b-7d2142271b61.png)
no i didn't miss it i added it so it doesn't work
